lambdas/listener/index.js

```diff
@@ -17,62 +16,58 @@ exports.handler = (event, context, callback) => {
   const productCode = message['product-code']
 
   switch (action) {
-    case 'subscribe-success':
-      subscribe(customerId, productCode, callback)
-      break
-    case 'subscribe-fail':
-      callback('not implemented')
-      break
-    case 'unsubscribe-pending':
-      callback('not implemented')
-      break
-    case 'unsubscribe-complete':
-      unsubscribe(customerId, productCode, callback)
-      break
+    case 'subscribe-success': return subscribe(customerId, productCode)
+    case 'subscribe-fail': throw new Error('not implemented')
+    case 'unsubscribe-pending': throw new Error('not implemented')
+    case 'unsubscribe-complete': return unsubscribe(customerId, productCode)
     default:
       console.log('Unknown action type ' + action)
-      callback('Invalid action: ' + action)
-      break
+      throw new Error('Invalid action: ' + action)
   }
 }
-/* eslint-enable standard/no-callback-literal */
 
-function subscribe (customerId, productCode, callback) {
+async function subscribe (customerId, productCode) {
   console.log(`Subscribing customer ${customerId} to product code ${productCode}`)
 
-  function err (err) {
-    console.log('error: ' + err)
-    callback(err)
-  }
-
-  // get identity id for marketplace customer id
-  controller.getCognitoIdentityId(customerId, err, identityId => {
+  try {
+    // get identity id for marketplace customer id
+    const identityId = await new Promise((resolve, reject) => {
+      controller.getCognitoIdentityId(customerId, reject, resolve)
+    })
     console.log('Got cognito identity : ' + identityId)
 
-    controller.getUsagePlanForProductCode(productCode, err, usagePlan => {
-      controller.subscribe(identityId, usagePlan.id, err, result => {
-        callback(null, result)
-      })
+    const usagePlan = await new Promise((resolve, reject) => {
+      controller.getUsagePlanForProductCode(customerId, reject, resolve)
     })
-  })
-}
-
-function unsubscribe (customerId, productCode, callback) {
-  console.log(`Unsubscribing customer ${customerId} from product code ${productCode}`)
 
-  function err (err) {
+    return await new Promise((resolve, reject) => {
```

Comment:
Why `return await`?

Reply:
In async functions, you have to `return await` if you want the exceptions to be caught correctly in surrounding `catch` blocks.

```js
// Logs `caught!`
try {
  return await Promise.reject(new Error())
} catch (err) {
  console.log("caught!")
}

// Doesn't log anything
try {
  return Promise.reject(new Error())
} catch (err) {
  console.log("caught!")
}
```

I blame TC39 for not having `return` implicitly `await` like literally everyone else intuitively expects it to so it can trigger `catch` blocks as appropriate.

Reply:
I see. I assumed the consuming function was catching b/c this try/catch rethrows, but now I see it is not.